### PR TITLE
[IDE] Refactoring: Expand a ternary operator into an if statement and vice-versa

### PR DESCRIPTION
**Pull request by romanroibu**
When selecting a ternary expression
```
let x = expression ? a : b
```
Suggest expanding to an if-else:
```
let x: Int
if expression {
    x = a
} else {
    x = b
}
```
And vice-versa.

Feedback is very much appreciated.

Resolves [SR-5794](https://bugs.swift.org/browse/SR-5794).

### CODE REVIEW

File: lib/IDE/Refactoring.cpp [hunk 1]
```diff
@@ -1667,6 +1667,366 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
   return false;
 }
 
+struct ExpandableTernaryExprInfo {
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: Can we make two sub-class of ExpandableTernaryExprInfo? Currently the logic for PatternBindingDecl and AssignExpr seems to be intertwined.

File: lib/IDE/Refactoring.cpp [hunk 2]
```diff
@@ -1667,6 +1667,366 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
...
+      && Info.Kind != RangeKind:: SingleExpression)
+    return notFound;
+
+  if (Info.ContainedNodes.size() != 1)
+    return notFound;
+
+  ExpandableTernaryExprInfo Target;
+
+  if (auto D = Info.ContainedNodes[0].dyn_cast<Decl*>())
+    Target.setBinding(dyn_cast<PatternBindingDecl>(D));
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: Once we separate the logic for PatternBindingDecl and AssignExpr, we can use constructors to pass the decl or the expr without using `setX()` functions.

File: lib/IDE/Refactoring.cpp [hunk 3]
```diff
@@ -1667,6 +1667,366 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
...
+}
+
+bool RefactoringActionExpandTernaryExpr::performChange() {
+  auto Target = findExpandableTernaryExpression(RangeInfo);
+
+  if (!Target.isValid())
+    return true; //abort
+
+  auto NameRange = Target.getNameRange();
+  auto NameCharRange = Lexer::getCharSourceRangeFromSourceRange(SM, NameRange);
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: can we simplify the logic by putting `Lexer::getCharSourceRangeFromSourceRange` into the `getNameRange` function.

File: lib/IDE/Refactoring.cpp [hunk 4]
```diff
@@ -1667,6 +1667,366 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
...
+  auto ThenCharRange = Lexer::getCharSourceRangeFromSourceRange(SM, ThenRange);
+
+  auto ElseRange = Target.getIf()->getElseExpr()->getSourceRange();
+  auto ElseCharRange = Lexer::getCharSourceRangeFromSourceRange(SM, ElseRange);
+
+  llvm::SmallString<64> DeclBuffer;
+  llvm::raw_svector_ostream OS(DeclBuffer);
+
+  std::string space(" ");
+  std::string newline("\n");
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: We can use `StringRef Space = " "` here. Also, It's good practice to have consistent name conventions, i.e. to have "Space" and "NewLine" Here.

**Review by nkcsgexi — COMMENTED** [review]
@romanroibu Overall looks good! 👍 I've reviewed the first half for now. I think it's a generally good idea to have one commit for each refactoring action though.

**Comment by romanroibu** [comment]
Thanks for the review, @nkcsgexi ! I'll commit the fixes and get back to you ;)

**Comment by romanroibu** [comment]
@nkcsgexi I've updated the PR with the changes you suggested. Let me know what else can be improved 👍

File: lib/IDE/Refactoring.cpp [hunk 5]
```diff
@@ -1667,6 +1667,391 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
   return false;
 }
 
+class ExpandableTernaryExprInfo {
+
+public:
+  virtual ~ExpandableTernaryExprInfo() {}
+
+  virtual swift::IfExpr *getIf() = 0;
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: We don't need `swift::` here. Could you clean them?

File: lib/IDE/Refactoring.cpp [hunk 6]
```diff
@@ -1667,6 +1667,391 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
...
+      return Invalid;
+
+    if (auto dest = Assign->getDest())
+      return dest->getSourceRange();
+
+    return Invalid;
+  }
+
+  swift::Type getType() {
+    return swift::Type();
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: `return nullptr;` should be sufficient here.

File: lib/IDE/Refactoring.cpp [hunk 7]
```diff
@@ -1667,6 +1667,391 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
...
+      return Pattern->getSourceRange();
+
+    return swift::SourceRange();
+  }
+
+  swift::Type getType() {
+    if (auto Pattern = getNamePattern())
+      return Pattern->getType();
+
+    return swift::Type();
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: `return nullptr;` here similarly.

File: lib/IDE/Refactoring.cpp [hunk 8]
```diff
@@ -1667,6 +1667,391 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
...
+      && Info.Kind != RangeKind:: SingleExpression)
+    return nullptr;
+
+  if (Info.ContainedNodes.size() != 1)
+    return nullptr;
+
+  if (auto D = Info.ContainedNodes[0].dyn_cast<Decl*>()) {
+    auto Binding = dyn_cast<PatternBindingDecl>(D);
+    auto BindingInfo = new ExpandableBindingTernaryExprInfo(Binding);
+    return std::unique_ptr<ExpandableTernaryExprInfo>(BindingInfo);
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: can we replace the above two statements with `return llvm::make_unique<ExpandableTernaryExprInfo>(Binding);`? In this way, we don't have a single point where a pointer is exposed.

File: lib/IDE/Refactoring.cpp [hunk 9]
```diff
@@ -1667,6 +1667,391 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
...
+
+  if (Info.Kind != RangeKind::SingleDecl
+      && Info.Kind != RangeKind:: SingleExpression)
+    return nullptr;
+
+  if (Info.ContainedNodes.size() != 1)
+    return nullptr;
+
+  if (auto D = Info.ContainedNodes[0].dyn_cast<Decl*>()) {
+    auto Binding = dyn_cast<PatternBindingDecl>(D);
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: Check if `Binding` is nullptr here and only create `ExpandableBindingTernaryExprInfo` when it is not. Therefore, we don't need to check whether `Binding` is null inside `ExpandableBindingTernaryExprInfo`.

File: lib/IDE/Refactoring.cpp [hunk 10]
```diff
@@ -1667,6 +1667,391 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
...
+    auto BindingInfo = new ExpandableBindingTernaryExprInfo(Binding);
+    return std::unique_ptr<ExpandableTernaryExprInfo>(BindingInfo);
+  }
+
+  if (auto E = Info.ContainedNodes[0].dyn_cast<Expr*>()) {
+    auto Assign = dyn_cast<AssignExpr>(E);
+    auto AssignInfo = new ExpandableAssignTernaryExprInfo(Assign);
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: Similarly, check if `Assign` is nullptr before creating `ExpandableAssignTernaryExprInfo`.

Comment (nkcsgexi) [reply lib/IDE/Refactoring.cpp]: e.g.
```
if (auto Assign = dyn_cast<AssignExpr>(E)) {
  return llvm::make_unique<>(...)...;
}
```

**Review by nkcsgexi — COMMENTED** [review]
Looks much better! Though, I've some new comments; could you take a look?

**Comment by romanroibu** [comment]
I've fixed the issues you pointed out 👍

**Comment by CodaFi** [comment]
Before this gets merged, could you squash this patch down into maybe two or three larger commits that encompass each feature that got added here?

**Comment by romanroibu** [comment]
Hey @nkcsgexi, I've addressed your latest comments, and also squashed the changes into 2 commits. Let me know if there's anything else that needs to be improved. Thanks!

**Comment by harlanhaskins** [comment]
This patch looks good to me as-is! Thanks for doing this! I just wanna get one more approval from @nkcsgexi and get one last CI pass.

@swift-ci please smoke test

**Review by nkcsgexi — APPROVED** [review]

**Comment by romanroibu** [comment]
@nkcsgexi I've fixed one issue with the tests, could you please trigger them again?

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by romanroibu** [comment]
I think all the issues should be solved, could you run the tests again, please?

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by romanroibu** [comment]
@nkcsgexi hmm, seems like it's failing because of the indentation. I thought the indentation is updated after the refactoring is applied, no? Should I just remove the indentation from the `.expected` files?

**Comment by nkcsgexi** [comment]
@romanroibu we do indent them when users are invoking refactorings through IDE. However, indentation is another operation orchestrated by source editor. In here, the expected change will only include changes from the refactoring, which doesn't include indentation : )

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@romanroibu Merged! Thank you for the awesome addition to Swift refactoring!

**Comment by romanroibu** [comment]
Thank you for all the help! :)

**Pull request merged**